### PR TITLE
feat(cdk): Default operation executor implementations

### PR DESCRIPTION
**Pull request by jdpgrailsdev**

### CODE REVIEW

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/util/ShutdownUtil.kt
Comment (postamar) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/util/ShutdownUtil.kt:51]: I'm worried that this approach isn't going to age well. Can we instead maintain a list of thread names that shouldn't be shut down and nuke anything not matching the list?

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt
Comment (postamar) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt:93]: Use `java.util.Scanner` along with `nextLine`? In any case don't roll your own.

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultReadOperationExecutor.kt
Comment (postamar) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultReadOperationExecutor.kt:65]: I'm not sure what Result.success(null) means. Is this `AirbyteMessage?` type all just to accommodate some edge case for check?

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultReadOperationExecutor.kt
Comment (postamar) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultReadOperationExecutor.kt:49]: Is `use` the right scoped method to use? I'm still not very familiar with these but I wouldn't expect `iterator` to be referenced directly in the body of whatever scoped method is appropriate (instead, relying on the implicit `this`).

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultCheckOperationExecutor.kt
Comment (postamar) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultCheckOperationExecutor.kt:23]: Why do we need both an Operation and an OperationExecutor? I'm sure there's a reason, but I don't get it.

Comment (jdpgrailsdev) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultCheckOperationExecutor.kt:23]: @postamar I attempted to answer this in an earlier comment, but it was a first pass/attempt at providing the appropriate logic split to allow for connectors to override the different behavior without needing to copy logic from the CDK.  For example, if you look at the CHECK operation, there is really two layers of logic:  the general logic that runs the check and if any exceptions are raised, returns a failure AND there is the actual check logic that talks to the source/destination system.  Most connectors will just want to override/replace the latter without having to touch the general logic.  There may be a few cases where the exception handling logic still needs to return success or something, so these two are separate out to facilitate that.  

That being said, I think that there may be three distinct operation types instead of one, so I am open to taking another pass at this as we implement some connectors using this new CDK.

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultReadOperationExecutor.kt
Comment (jdpgrailsdev) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultReadOperationExecutor.kt:49]: @postamar This is generated code from the Java -> Kotlin conversion, so I will take another pass to see if I can clean this up.

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultReadOperationExecutor.kt
Comment (jdpgrailsdev) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultReadOperationExecutor.kt:65]: @postamar This highlights the fact that there are really three different operations:  one that returns a single airbyte message (check, spec, discover), one that returns nothing (write) and one that returns a list of airbyte message.  Currently, there is just one interface, so the common denominator is to just return a possible single airbyte message.  This can be improved upon once we get this in place.

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultReadOperationExecutor.kt
Comment (jdpgrailsdev) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultReadOperationExecutor.kt:49]: @postamar I take this back.  The use of `use` is to ensure that the iterator is automatically closed when it is done being used.

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt
Comment (gisripa) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt:93]: Probably there is a subtle difference. Bis.read blocks when we get records in bursts, also we use a bounded byteArray buffer, scanner.nextLine will continue to buffer until it finds the linefeed char.  also the downstream call relies in bufferedOutStream.size to allocate/block consumption.

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/util/ShutdownUtil.kt
Comment (jdpgrailsdev) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/util/ShutdownUtil.kt:51]: @postamar I'll create a ticket for this.  This is a port of the existing logic.  I have tried to not touch things that are not part of the main path to getting this change in, as it is already taking weeks to get what should be a simple, net-new first pass merged.  This feels like something that the team should take on as technical debt paydown.

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultCheckOperationExecutor.kt
Comment (postamar) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultCheckOperationExecutor.kt:23]: Got it, thanks!

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt
Comment (postamar) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt:93]: OK, still, let's not merge this. Either it's just bad code, or it's deliberately and carefully designed in this way but it's lacking commentary to justify its unique special snowflake status by not using a Scanner or an InputStreamReader.

Personally I find the whole memory considerations completely ridiculous in light of the use of `String` and besides it's not like we're anywhere close to supporting gigabyte-sized records anyway.

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/util/ShutdownUtil.kt
Comment (postamar) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/util/ShutdownUtil.kt:51]: Yeah let's defer this in that case. Thanks for creating the ticket.

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt
Comment (jdpgrailsdev) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt:93]: @postamar This is the existing code from the Java implementation of the `IntegrationRunner` ported to Kotlin.  I did not change or introduce this.  This should not hold up the PR.  If you want to take a stab at fixing it, I suggest we do that as a follow on PR as this is pre-existing code.

Comment (postamar) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt:93]: https://github.com/airbytehq/airbyte/pull/36375
Was this really worth arguing over and deferring? I'm puzzled.

Comment (gisripa) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt:93]: @postamar It is definitely lacking commentary (predates me) but I found an interesting test which was specifically testing the output of this method against `java.util.Scanner.in..` as you did in your PR as `oldoutput`.  
https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/java/airbyte-cdk/core/src/test/java/io/airbyte/cdk/integrations/base/IntegrationRunnerBackwardsCompatbilityTest.java#L36 I would prefer this as a follow up too with specific

Comment (jdpgrailsdev) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt:93]: > #36375 Was this really worth arguing over and deferring? I'm puzzled.

@postamar In normal circumstances no.  However, this PR and its predecessor have been open for over two weeks.  If we try to fix every ill that exists in the current code that is either a) being converted to Kotlin or b) being moved into the new framework, this will never merge.  I am totally happy to address these as part of a punch list once this gets merged.  Thank you for taking some time to fix this particular issue.

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultReadOperationExecutor.kt
Comment (stephane-airbyte) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultReadOperationExecutor.kt:59]: why the need to wrap an IllegalArgumentException into an OperationExecutionException here?

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt
Comment (stephane-airbyte) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt:93]: can we at least share implementation rather than copy-pasting code? Now that we've converted everything to kotlin, I don't think there's any reason not to do that

Comment (gisripa) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt:93]: @stephane-airbyte  Joe has one in [flight](https://github.com/airbytehq/airbyte/pull/36673) changing this. so shouldn't be a blocker for this PR maybe ?

Comment (jdpgrailsdev) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt:93]: @stephane-airbyte @gisripa We should probably wait for @jbfbell 's PR and then do a refactor to extract this logic to a common location that we can use in both places.

Comment (gisripa) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt:93]: Sounds good to me. Oh I see, we want to share this logic between old IntegrationRunner and the new µnaut runner. 👍

Comment (jbfbell) [reply airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/operation/executor/DefaultWriteOperationExecutor.kt]: chiming in here that my PR is merged

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/util/WriteStreamConsumer.kt
Comment (jbfbell) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/util/WriteStreamConsumer.kt:41]: super unfair comment that is not blocking - I was thinking about this last night and these conditions should probably be reversed since once its above 10k I didn't want to keep logging

**Review by jbfbell — APPROVED** [review]

**Review by gisripa — APPROVED** [review]

**Pull request closed without merging**
